### ISSUE TITLE
JsonResponse::setData with JSON_THROW_ON_ERROR can throw when there is no error

### ISSUE THREAD

**Issue by lt** [symfony/symfony#31447]
**Symfony version(s) affected**: 4.2.5

**Description**  

`JsonResponse::setData` checks `json_last_error` to determine whether or not encoding was successful, however when `JSON_THROW_ON_ERROR` is set as an encoding option, the error state is not cleared.

This means when a previous JSON operation that does not use `JSON_THROW_ON_ERROR` has failed, `JsonResponse::setData` will throw an exception even when the encode operation was successful.

**How to reproduce**  

```
<?php declare(strict_types=1);

require 'vendor/autoload.php';

$response = new \Symfony\Component\HttpFoundation\JsonResponse();
$response->setEncodingOptions(JSON_THROW_ON_ERROR);

// Commenting this out prevents the exception from setData
json_decode('bad data, but carry on');

$response->setData('this is fine');
```
```
PHP Fatal error:  Uncaught InvalidArgumentException: Syntax error in /var/www/html/vendor/symfony/http-foundation/JsonResponse.php:152                                                                                                       
Stack trace:
#0 /var/www/html/test.php(12): Symfony\Component\HttpFoundation\JsonResponse->setData('"this is fine"')
#1 {main}
  thrown in /var/www/html/vendor/symfony/http-foundation/JsonResponse.php on line 152
```
**Possible Solution**  

Check encoding options for `JSON_THROW_ON_ERROR` when checking `json_last_error()`

**Additional context**  

https://bugs.php.net/bug.php?id=77997

**Comment by nicolas-grekas** [comment]
Related to https://externals.io/message/105653

**Comment by jojotjebaby** [comment]
I am currently looking into this issue and am not sure we can prevent this behaviour because it would be impossible to know if the user did a json_(encode / decode) before calling setData with JSON_THROW_ERROR as encodingOption.

Even if the user set the encodingOption to JSON_THROW_ERROR and we added a check before doing json_last_error(), we cannot now which json function has thrown the error. Another problem is that the try catch only works for php7.3 and we have to use json_last_error() for php7.2 and php7.1

Any suggestions on what to do with this issue ?

**Comment by JJarrie** [comment]
What about just fixing by internal flag's checking? Just to have explicit errors catching in the component, waiting the discussion about the behavior of the global error state be finished.

```php
        if (!($this->encodingOptions & 4194304) && JSON_ERROR_NONE !== json_last_error()) {
            throw new \InvalidArgumentException(json_last_error_msg());
        }

```

**Comment by nicolas-grekas** [comment]
Please check #31860

**Comment by nicolas-grekas** [comment]
Better fix in #31869
